roles/swlib/tasks/gcscopy.yml

```diff
@@ -98,21 +98,8 @@
 - name: gcscopy | Copy GI OPatch update files from GCS to target instance
```

Comment:
I think there might be a misunderstanding of what "gcscopy" means;  it was specifically designed to avoid dependencies on network connectivity and authentication state on the BMS host, and instead to rely on the control node's default auth as a cloud VM.

Reply:
The commit message for this change details the situations where this could be a problem, which indeed I faced with patching failing as the latest opatch in gcs bucket was not getting over to the bms host.

Reply:
Correction here:  this code has a "delegate_to: 127.0.0.1", so actually runs on the control node itself.

This same logic is on line 16 and line 59.  It looks like there is an attempt to error checking, as well as unzipping the patchfile.  From observation, it looks like we conditionally check if the patch file exists already on the BMS host, and if its integrity is OK.  (seemingly overlapping line 123's behavior).

I'm guessing that you're hitting a case of partial swlib corruption that these two checks didn't cover.  So clearing out swlib manually would probably fix your immediate issue.

From a toolkit perspective though, I'd be OK if we unconditionally copied every time, given that the use case is to install Oracle which should typically happen once.  (A fancier solution might use gsutil rsync over sshfs, but probably overkill).  I'd just ask that it be done consistently for each of the calls in this playbook.  (And potentially as a separate PR).

Reply:
Recap-ing the existing logic as follows:
1. look for the existence of the patch file about to be copied over from gcs
2. if it exists 
   a.  and if it's a valid zip file, do nothing
   b. but if zip file intergrity checks fails, copy over from gcs
3. if it doesn't exist, copy over from gcs

While this is a good logic for any Oracle patch files, as patch numbers are unique, the only failing case for this is the opatch file name. The opatch file name is always the same for a given platform and a give version, example: `p6880880_190000_Linux-x86-64.zip` or `p6880880_122010_Linux-x86-64.zip`.

Hence the change in this PR only impacted the opatch copy from gcs and leave the rest of the existing logic for other file copy untouched.

The error that prompted this change is:
```
TASK [opatch_apply | Run opatch apply] *****************************************************************************************************************************************
fatal: [linuxserver44.orcl]: FAILED! => {"changed": true, "cmd": ["/u01/app/19.3.0/grid/OPatch/opatchauto", "apply"], "delta": "0:00:33.136668", "end": "2021-06-09 23:30:08.443024", "failed_when_result": true, "msg": "non-zero return code", "rc": 42, "start": "2021-06-09 23:29:35.306356", "stderr": "", "stderr_lines": [], "stdout": "
OPatchauto session is initiated at Wed Jun  9 23:29:35 2021
...
Patch: /u01/oracle_install/32126842/32226239/32218454
Log: /u01/app/oracle/product/19.3.0/dbhome_1/cfgtoollogs/opatchauto/core/opatch/opatch2021-06-09_23-29-49PM_1.log
Reason: Failed during Analysis: CheckMinimumOPatchVersion Failed, [ Prerequisite Status: FAILED, Prerequisite output: 
The details are:


The OPatch being used has version 12.2.0.1.21 while the following patch(es) require higher versions: 
Patch 32218454 requires OPatch version 12.2.0.1.23.
Please download latest OPatch from My Oracle Support.
] 
```

In the interest of efficiency of gcscopy.yml from a time tken perspective, it may be better to retain existing logic where the file names are unique and the zip file checks come out as passed.

Reply:
Unlike other Oracle patches, opatch patch file (*.zip) names have the same as the previous release.
In other words, opatch updates are packaged inside of the same filename from Oracle Support download.

Idea is that as the file's name is the same, and it `/u01/oracle_install` (or the staging area) has the same file as follows:
```[oracle@at-00010-svr002 oracle_install]$ ll -h
...
-rw-rw-r-- 1 ansible ansible  117M Jul  1 11:37 p6880880_190000_Linux-x86-64.zip
```

then existing logic skips copying the "newer" p6880880_190000_Linux-x86-64.zip from gcs bucket.
Hence the change.

Reply:
So as I gather, the zipfile name for a given patch is unique, with the exception for OPatch itself, where a point release can have the same zipfile as the previous one.  So if a previous release is cached in {{ swlib_path }} and a newer one becomes available, we can't invalidate the patch.  This change downloads OPatch from the GCS bucket every time;  the overhead should be minimal as OPatch itself is quite small.

Can we add a comment to this effect in the code, for those confused about this inconsistent behavior?

Reply:
Adding the following as comment above the block:
```
# overwrite opatch zipfile on BMS host from gcs every time that may be cached currently in {{ swlib_path }}
# reason to overwrite is that the opatch zipfile in MOS is named the same, example: p6880880_190000_Linux-x86-64.zip
# this change downloads OPatch from GCS bucket every time with neglible overhead as opatch zip file is typically ~120M
```